mirgecom/profiling.py

```diff
@@ -138,48 +92,113 @@ def __init__(self, queue, allocator=None) -> None:
+        evt = knl(queue, gs, ls, *actual_args, wait_for=wait_for)
+
+        name = knl.function_name
+        knl.name = knl.function_name  # "name" is used by the LoopKernels
+
+        args_tuple = tuple(
+            (arg.size)
+            for arg in actual_args if isinstance(arg, cl.array.Array))
+
+        try:
+            ignore = self.kernel_stats[knl][args_tuple]  # noqa
+        except KeyError:
+            nbytes = 0
+            nops = 0
+            for arg in actual_args:
+                if isinstance(arg, cl.array.Array):
+                    nbytes += arg.size * arg.dtype.itemsize
+                    nops += arg.size
+            res = SingleCallKernelProfile(time=0, flops=nops, bytes_accessed=nbytes,
+                                footprint_bytes=nbytes)
+            self.kernel_stats.setdefault(knl, {})[args_tuple] = res
+
+        if self.logmgr and f"{name}_time" not in self.logmgr.quantity_data:
+            self.logmgr.add_quantity(KernelProfile(self, name))
+
+        self.profile_events.append(ProfileEvent(evt, knl, args_tuple))
+
+        return evt
 
     def _finish_profile_events(self) -> None:
         # First, wait for completion of all events
         if self.profile_events:
             cl.wait_for_events([pevt.cl_event for pevt in self.profile_events])
 
-        global nonloopy_profile_events
-        if nonloopy_profile_events:
-            cl.wait_for_events([npe.cl_event for npe in nonloopy_profile_events])
-
         # Then, collect all events and store them
         for t in self.profile_events:
             program = t.program
             r = self._get_kernel_stats(program, t.args_tuple)
             time = t.cl_event.profile.end - t.cl_event.profile.start
 
-            new = ProfileResult(time, r.flops, r.bytes_accessed, r.footprint_bytes)
-
-            self.profile_results.setdefault(program, []).append(new)
+            new = SingleCallKernelProfile(time, r.flops, r.bytes_accessed,
+                                          r.footprint_bytes)
 
-        for t in nonloopy_profile_events:
-            program = t.knl
-            time = t.cl_event.profile.end - t.cl_event.profile.start
-            new = ProfileResult(time, t.nops, t.nbytes, t.nbytes)
             self.profile_results.setdefault(program, []).append(new)
 
         self.profile_events = []
-        nonloopy_profile_events = []
 
-    def tabulate_profiling_data(self) -> pytools.Table:
+    def get_and_reset_profiling_data_for_kernel(self, kernel_name: str,
+                                 wait_for_events=True) -> MultiCallKernelProfile:
+        """Return and reset profiling data for kernel `kernel_name`."""
+        if wait_for_events:
+            self._finish_profile_events()
+
+        num_calls = 0
+        time = 0
+        flops = 0
+        bytes_accessed = 0
+        fprint_bytes = 0
+
+        keys = [k for k in self.profile_results.keys() if k.name == kernel_name]
```

Comment:
Continued from https://github.com/illinois-ceesd/mirgecom/pull/96#discussion_r578866021:

This isn't better IMO; if anything, it's slower because the list of kernel-things it traverses is longer. Lookup by name should be O(1), otherwise you're using the wrong data structure.

I looked, I wasn't able to find any code that uses anything other than the kernel name. Could you point me to it?

Reply:
> Continued from [#96 (comment)](https://github.com/illinois-ceesd/mirgecom/pull/96#discussion_r578866021):
> 
> This isn't better IMO; if anything, it's slower because the list of kernel-things it traverses is longer. Lookup by name should be O(1), otherwise you're using the wrong data structure.

`len(self.profile_results.keys())` should be pretty short, say O(10). `len(keys)` should be in the worst case O(3). The current implementation is about 20% faster of what I had before.

> I looked, I wasn't able to find any code that uses anything other than the kernel name. Could you point me to it?

It's used in `_get_profiling_stats()` directly:

https://github.com/illinois-ceesd/mirgecom/blob/ce51d324754b94e26b0783dd635cd97d99bdcc69/mirgecom/profiling.py#L267-L269

and then the result of that is used indirectly in `tabulate_profiling_data()`:

https://github.com/illinois-ceesd/mirgecom/blob/ce51d324754b94e26b0783dd635cd97d99bdcc69/mirgecom/profiling.py#L201

 `tabulate_profiling_data()`, in contrast to the logging of kernel profiling data, *does* differentiate between different kernels and kernel arguments when presenting results. So we could invert the way results are stored and index them by kernel name, speeding up the logging while slowing down `tabulate_profiling_data()` (or just not differentiating between kernels of the same name). Imho, that would be a fair tradeoff, but changing this might be a big enough change for a new PR.

Reply:
Can you explain the benefit of differentiating beyond the name? (/provide an example of that differentiation) I still don't really understand what you mean here.

Reply:
> Can you explain the benefit of differentiating beyond the name? (/provide an example of that differentiation) I still don't really understand what you mean here.

I think there are two main reasons:

1. In practical terms, it means that kernels could be analyzed separately based on their input parameters, e.g. in the following profiling output for the `grudge_assign_0` kernel: 

Function            | Calls | Time_sum [s] | Time_min [s] | Time_avg [s] | Time_max [s] | GFlops/s_min | GFlops/s_avg | GFlops/s_max | BWAcc_min [GByte/s] | BWAcc_mean [GByte/s] | BWAcc_max [GByte/s] | BWFoot_min [GByte/s] | BWFoot_mean [GByte/s] | BWFoot_max [GByte/s] | Intensity (flops/byte)
:-------------------|:------|:-------------|:-------------|:-------------|:-------------|:-------------|:-------------|:-------------|:--------------------|:---------------------|:--------------------|:---------------------|:----------------------|:---------------------|:-----------------------
grudge_assign_0     | 2     | 0.0001233    | 4.629e-05    | 6.163e-05    | 7.696e-05    | 1.029        | 1.37         | 1.711        | 6.081               | 8.096                | 10.11               | 0.00018              | 0.00018               | 0.00018              | 0.1692
grudge_assign_0     | 8     | 0.0004143    | 3.341e-05    | 5.179e-05    | 6.104e-05    | 0.118        | 0.1444       | 0.2155       | 1.769               | 2.166                | 3.233               | 0.000108             | 0.000108              | 0.000108             | 0.06667
grudge_assign_0     | 4     | 0.000175     | 3.676e-05    | 4.376e-05    | 5.955e-05    | 0.1209       | 0.1707       | 0.1958       | 1.814               | 2.56                 | 2.938               | 0.000108             | 0.000108

If I remember correctly, this was a feature you had requested.

2. Differentiating by name could lead to confusion in case two different kernels happen to have the same name (like `diff` in grudge some days ago).

Imho, neither of these two reasons is compelling enough to keep the current system in place. Apart from the performance, there is also a slight annoyance of the output not matching the e.g. nvprof output in terms of the number of different kernels.

Reply:
> analyzed separately based on their input parameters

I'm still not sure I understand. How are these differentiated? By the shape of the arguments? By the pointers? By the actual values in the array?

Reply:
> > analyzed separately based on their input parameters
> 
> I'm still not sure I understand. How are these differentiated? By the shape of the arguments? By the pointers? By the actual values in the array?

Input parameters are differentiated by the shape of array args and values of scalar args. See e.g.

https://github.com/illinois-ceesd/mirgecom/blob/785efc06c40da0a1ba9b708248dc3f9211ca7be8/mirgecom/profiling.py#L292-L298

This is then used in the result collection:

https://github.com/illinois-ceesd/mirgecom/blob/785efc06c40da0a1ba9b708248dc3f9211ca7be8/mirgecom/profiling.py#L143-L151

Reply:
I see. Thanks! I guess I vaguely remember pushing for this, but on second thought I'm no longer fond of this. If there are important distinctions between kernels, then we can probably just encode them in the name. Seeing the "same" kernel listed five times (but not really knowing which is which) isn't all that helpful after all. How about name-only after all?

Reply:
Changed to indexing by name only. Did you see my other question regarding the deletion @inducer?

> How would the "reset" get called from the logging infrastructure?

Reply:
> How would the "reset" get called from the logging infrastructure?

Likely the `LogQuantity` would reset the stats right after reading them. So pretty similar to before, just not forcibly connected.

Reply:
Ok, I have implemented that.